std.math: add nextAfter

castholm opened the pull request:
`nextAfter()` returns the next representable value after `x` in the direction of `y` and is a standard math library function ([C++](https://en.cppreference.com/w/cpp/numeric/math/nextafter), [Java](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#nextAfter-double-double-)). It is primarily useful for bitwise incrementing/decrementing floats.

This implementation supports runtime integers, runtime floats and `comptime_int`. `comptime_float` is not supported because NaNs/infinities are intentionally difficult to obtain and because I'm not sure if the fact that it's backed by `f128` is supposed to be an implementation detail. Either way, the user could just call the function with the floating-point type whose behavior they want at comptime and then cast the result to `comptime_float`.

The float implementation was ported from mingw-w64 with some slight changes made possible because the Zig standard library doesn't care about raising FP exceptions.

The number of test cases may seem excessive but they should cover every normal and edge case for every float type and are especially important for verifying that `f80` works.

castholm commented:
Failing builds should hopefully pass once #16900 is merged unless there are more subtle `f80` issues.

matu3ba commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+        try expect(nextAfter(T, -math.inf(T), -math.inf(T)) == -math.inf(T));
+        try expect(nextAfter(T, -math.inf(T), math.inf(T)) == -math.floatMax(T));
+        try expect(nextAfter(T, -math.floatMax(T), -math.inf(T)) == -math.inf(T));
+        try expect(math.isNan(nextAfter(T, 1.0, math.nan(T))));
```
To make things complete, add +-inf -> nan and vice versa.

castholm replied:
Added.

matu3ba commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+        try expect(nextAfter(comptime_int, -3, 7) == -2);
+        try expect(nextAfter(comptime_int, 5, 5) == 5);
+        try expect(nextAfter(comptime_int, -5, -5) == -5);
+        try expect(nextAfter(comptime_int, 0, 0) == -0);
```
reasonably large min/max comptime value

matu3ba commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+    try expect(nextAfter(u1, 0, 1) == 1);
+    try expect(nextAfter(u1, 1, 1) == 1);
+    try expect(nextAfter(u1, 1, 0) == 0);
+    inline for (.{ i8, i16, i32, i64, i128, i333 }) |T| {
```
why not min and max unsigned and signed integers Zig supports?  u/i65535

castholm replied:
`i/u65535` blows up compile times and memory consumption to a ridiculous level, at least it did for me when I tried using them. I included `i/u333` to have arbitrary bit-width ints represented.

matu3ba commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+const assert = std.debug.assert;
+const expect = std.testing.expect;
+
+/// Returns the next representable value after `x` in the direction of `y`.
```
Not sure, if representable is ideal as some float values are not representable NaN, +-inf and stepping into NaN is NaN. Also sNaN is not supported, but that is an overall LLVM issue.

castholm replied:
Most documentation for C/C++ `nextafter()` seem to use "next representable (floating-point) value" so I personally think it should be clear enough. We could update it to clarify edge cases:

> Returns the next representable value after `x` in the direction of `y`.
>
> Special cases:
> * If `x == y`, `y` is returned.
> * For floats, if either `x` or `y` is a NaN, a NaN is returned.
> * For floats, if `x == 0.0` and `@abs(y) > 0.0`, the smallest subnormal number with the sign of `y` is returned.

matu3ba reviewed (changes requested):
A bit more test coverage would be good and nit on function doc comment.

castholm commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+    return if (@typeInfo(T) == .Int and @typeInfo(T).Int.bits < 2)
+        // Special case for `i0`, `u0`, `i1`, and `u1`.
+        y
+    else if (y > x)
```
I'm not an assembly person but this seems to compile into branchless code.

```asm
nextAfter_i32:
        lea     eax, [rdi - 1]
        cmp     esi, edi
        lea     ecx, [rdi + 1]
        cmovge  eax, esi
        cmovg   eax, ecx
        ret

nextAfter_u64:
        lea     rax, [rdi - 1]
        cmp     rsi, rdi
        lea     rcx, [rdi + 1]
        cmovae  rax, rsi
        cmova   rax, rcx
        ret
```
https://godbolt.org/z/qhx1cPqn4

castholm commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,329 @@
+    try expect(nextAfter(f128, -0x0.FFFFFFFFFFFFFFFFFFFFFFFFFFFFp-16382, -math.inf(f128)) == -0x1p-16382);
+
+    // special values
+    inline for (.{ f16, f32, f64, f128 }) |T| {
```
`nextAfter(T, math.inf(T), -math.inf(T)) == math.floatMax(T)` and variants were already included.

castholm commented:
Moved the `f80` test cases back with the rest of the float types now that the underlying issues seem resolved. I also updated the doc comment to include special cases and added some missing useful infinity/NaN test cases.

castholm commented on lib/std/math/nextafter.zig:
```diff
@@ -0,0 +1,304 @@
+    if (math.isNan(x) or math.isNan(y)) {
+        return math.nan(T);
+    }
+    if (x == 0.0) {
+        const sign_bit_mask = @as(Bits, 1) << (@bitSizeOf(T) - 1);
+        const smallest_bits: Bits = @bitCast(math.floatTrueMin(T));
+        const y_sign_bit = @as(Bits, @bitCast(y)) & sign_bit_mask;
+        return @bitCast(smallest_bits | y_sign_bit);
+    }
```
I can think of at least three ways of returning +-smallest here but I don't know which way is the best and easiest for the compiler to optimize.

```zig
const std = @import("std");
const math = std.math;

export fn copySign(y: f32) f32 {
    const sign_bit_mask: u32 = 1 << (@bitSizeOf(f32) - 1);
    const smallest_bits: u32 = @bitCast(math.floatTrueMin(f32));
    const y_sign_bit = @as(u32, @bitCast(y)) & sign_bit_mask;
    return @bitCast(smallest_bits | y_sign_bit);
}

export fn compareFloat(y: f32) f32 {
    return if (y > 0.0)
        math.floatTrueMin(f32)
    else
        -math.floatTrueMin(f32);
}

export fn compareSignedInt(y: f32) f32 {
    const y_bits_signed: i32 = @bitCast(y);
    return if (y_bits_signed >= 0)
        math.floatTrueMin(f32)
    else
        -math.floatTrueMin(f32);
}
```

```asm
copySign:
        vmovmskps       eax, xmm0
        shl     eax, 31
        or      eax, 1
        vmovd   xmm0, eax
        ret

.LCPI1_0:
        .long   0x80000001
.LCPI1_1:
        .long   0x00000001
compareFloat:
        vxorps  xmm1, xmm1, xmm1
        vcmpltss        k1, xmm1, xmm0
        vmovss  xmm0, dword ptr [rip + .LCPI1_0]
        vmovss  xmm0 {k1}, dword ptr [rip + .LCPI1_1]
        ret

.LCPI2_0:
        .long   0x00000001
.LCPI2_1:
        .long   0x80000000
compareSignedInt:
        vpbroadcastd    xmm1, dword ptr [rip + .LCPI2_0]
        vpternlogd      xmm0, xmm1, dword ptr [rip + .LCPI2_1]{1to4}, 236
        ret
```

https://godbolt.org/z/x6zevjd4T

castholm replied:
I thought about it some more and decided to change this into the `if (y > 0.0) trueMin else -trueMin` variant. All three options are branchless (which is probably the most important thing) and out of the three it's the most immediately obvious and readable one.

castholm commented:
Replaced the `@bitCast`s from/to `f80` with `math.break/break_f80`.

castholm commented:
Added `c_longdouble` test cases.

andrewrk commented:
Nice work! I hope you don't mind, I pasted your excellent PR writeup into the commit message upon merging.
